File: idscp2/src/main/java/de/fhg/aisec/ids/idscp2/idscp_core/finite_state_machine/Timer.java
package de.fhg.aisec.ids.idscp2.idscp_core.finite_state_machine;

import java.util.concurrent.locks.ReentrantLock;

/**
 * A Timer class that provides an API to the FSN to start and cancel timeout threads
 * The timer ensures that no canceled timer is able to trigger a timeout transitions
 *
 * @author Leon Beckmann (leon.beckmann@aisec.fraunhofer.de)
 */
public class Timer {

    private TimerThread thread = null;
    private final ReentrantLock fsmIsBusy;
    private final ReentrantLock mutex = new ReentrantLock(true);
    private final Runnable timeoutHandler;

    Timer(ReentrantLock fsmIsBusy, Runnable timeoutHandler) {
        this.fsmIsBusy = fsmIsBusy;
        this.timeoutHandler = timeoutHandler;
    }

    void resetTimeout(long delay) {
        cancelTimeout();
        start(delay);
    }

    /*
     * Start a timer thread that triggers the timeout handler routine after a given timout delay
     */
    public void start(long delay) {
        mutex.lock();
        thread = new TimerThread(delay, timeoutHandler, fsmIsBusy);
        thread.start();
        mutex.unlock();
    }

    /*
     * Cancel the current timer thread
     */
    void cancelTimeout() {
        mutex.lock();
        if (thread != null) {
            thread.safeStop();
            thread = null;
        }
        mutex.unlock();
    }
}

File: idscp2/src/main/kotlin/de/fhg/aisec/ids/idscp2/idscp_core/fsm/Timer.kt
package de.fhg.aisec.ids.idscp2.idscp_core.fsm

import java.util.concurrent.locks.ReentrantLock

/**
 * A Timer class that provides an API to the FSN to start and cancel timeout threads
 * The timer ensures that no canceled timer is able to trigger a timeout transitions
 *
 * @author Leon Beckmann (leon.beckmann@aisec.fraunhofer.de)
 */
class Timer internal constructor(private val fsmIsBusy: ReentrantLock, private val timeoutHandler: Runnable) {
    private var thread: TimerThread? = null
    private val mutex = ReentrantLock(true)
    fun resetTimeout(delay: Long) {
        cancelTimeout()
        start(delay)
    }

    /*
     * Start a timer thread that triggers the timeout handler routine after a given timout delay
     */
    fun start(delay: Long) {
        mutex.lock()
        thread = TimerThread(delay, timeoutHandler, fsmIsBusy)
        thread!!.start()
        mutex.unlock()
    }

    /*
     * Cancel the current timer thread
     */
    fun cancelTimeout() {
        mutex.lock()
        if (thread != null) {
            thread!!.safeStop()
            thread = null
        }
        mutex.unlock()
    }
}